dd-sdk-android/src/main/java/com/datadog/opentracing/propagation/B3MHttpCodec.java

```diff
@@ -0,0 +1,143 @@
+  private static final String SAMPLING_PRIORITY_DROP = String.valueOf(0);
+  private static final int HEX_RADIX = 16;
+
+  private B3MHttpCodec() {
+    // This class should not be created. This also makes code coverage checks happy.
+  }
+
+  public static class Injector implements HttpCodec.Injector {
+
+    @Override
+    public void inject(final DDSpanContext context, final TextMapInject carrier) {
+      try {
+        carrier.put(TRACE_ID_KEY, context.getTraceId().toString(HEX_RADIX).toLowerCase(Locale.US));
+        carrier.put(SPAN_ID_KEY, context.getSpanId().toString(HEX_RADIX).toLowerCase(Locale.US));
+
+        if (context.lockSamplingPriority()) {
+          carrier.put(
+              SAMPLING_PRIORITY_KEY, convertSamplingPriority(context.getSamplingPriority()));
+        }
+      } catch (final NumberFormatException e) {
+      }
+    }
+
+    private String convertSamplingPriority(final int samplingPriority) {
+      return samplingPriority > 0 ? SAMPLING_PRIORITY_ACCEPT : SAMPLING_PRIORITY_DROP;
+    }
+  }
+
+  public static class Extractor implements HttpCodec.Extractor {
+
+    private final Map<String, String> taggedHeaders;
+
+    public Extractor(final Map<String, String> taggedHeaders) {
+      this.taggedHeaders = new HashMap<>();
+      for (final Map.Entry<String, String> mapping : taggedHeaders.entrySet()) {
+        this.taggedHeaders.put(mapping.getKey().trim().toLowerCase(Locale.US), mapping.getValue());
+      }
+    }
+
+    @Override
+    public SpanContext extract(final TextMapExtract carrier) {
+      try {
+        Map<String, String> tags = Collections.emptyMap();
+        BigInteger traceId = BigInteger.ZERO;
+        BigInteger spanId = BigInteger.ZERO;
+        int samplingPriority = PrioritySampling.UNSET;
+
+        for (final Map.Entry<String, String> entry : carrier) {
+          final String key = entry.getKey().toLowerCase(Locale.US);
+          final String value = entry.getValue();
+
+          if (value == null) {
+            continue;
+          }
+
+          if (TRACE_ID_KEY.equalsIgnoreCase(key)) {
+            final String trimmedValue;
+            final int length = value.length();
+            if (length > 32) {
+              traceId = BigInteger.ZERO;
+              continue;
+            } else if (length > 16) {
+              trimmedValue = value.substring(length - 16);
+            } else {
+              trimmedValue = value;
+            }
+            traceId = validateUInt64BitsID(trimmedValue, HEX_RADIX);
+          } else if (SPAN_ID_KEY.equalsIgnoreCase(key)) {
+            spanId = validateUInt64BitsID(value, HEX_RADIX);
+          } else if (SAMPLING_PRIORITY_KEY.equalsIgnoreCase(key)) {
+            samplingPriority = convertSamplingPriority(value);
+          }
+
+          if (taggedHeaders.containsKey(key)) {
+            if (tags.isEmpty()) {
+              tags = new HashMap<>();
+            }
```

Comment:
No because in Java `Collections.emptyMap()` returns an immutable map. Here we change it to a new object so we can then add data to it. 

But bear in mind this file was originally the B3M encoder/decoder that already existed in the `DDTrace.java` lib.

Reply:
indeed, didn't notice it is `emptyMap` by default.